src/vmm/src/devices/virtio/balloon/device.rs

```diff
@@ -1412,16 +1509,252 @@ pub(crate) mod tests {
         }
     }
 
+    #[test]
+    fn test_process_reporting() {
+        let mem = create_virtio_mem();
+        let mut th =
+            VirtioTestHelper::<Balloon>::new(&mem, Balloon::new(0, true, 0, false, true).unwrap());
+
+        th.activate_device(&mem);
+
+        let page_size = host_page_size() as u64;
+
+        // This has to be u32 for the scatter gather
+        #[allow(clippy::cast_possible_truncation)]
+        let page_size_chain = page_size as u32;
+        let reporting_idx = th.device().free_page_reporting_idx();
+
+        let safe_addr = align_up(th.data_address(), page_size);
+
+        th.add_scatter_gather(reporting_idx, 0, &[(0, safe_addr, page_size_chain, 0)]);
+        check_metric_after_block!(
+            METRICS.free_page_report_freed,
+            page_size,
+            invoke_handler_for_queue_event(&mut th.device(), reporting_idx)
+        );
+
+        // Test with multiple items
+        th.add_scatter_gather(
+            reporting_idx,
+            0,
+            &[
+                (0, safe_addr, page_size_chain, 0),
+                (1, safe_addr + page_size, page_size_chain, 0),
+                (2, safe_addr + (page_size * 2), page_size_chain, 0),
+            ],
+        );
+
+        check_metric_after_block!(
+            METRICS.free_page_report_freed,
+            page_size * 3,
+            invoke_handler_for_queue_event(&mut th.device(), reporting_idx)
+        );
+
+        // Test with unaligned length
+        th.add_scatter_gather(reporting_idx, 0, &[(1, safe_addr + 1, page_size_chain, 0)]);
+
+        check_metric_after_block!(
+            METRICS.free_page_report_fails,
+            1,
+            invoke_handler_for_queue_event(&mut th.device(), reporting_idx)
+        );
+    }
+
+    #[test]
+    fn test_process_hinting() {
```

Comment:
nit: consider splitting this test into multiple self-contained ones, each testing its own scenario.

Reply:
Will do thanks!